api/src/services/setup/upgrade-log.js

```diff
@@ -0,0 +1,231 @@
+const db = require('../../db');
+const logger = require('../../logger');
+
+/**
+ * @typedef {Object} UpgradeLog
+ * @property {string} _id
+ * @property {string} user - username of user initiating the upgrade (eg. "mary")
+ * @property {string} from_version - eg "4.0.3"
+ * @property {string} to_version - eg "4.1.2"
+ * @property {number} start_date - timestamp when upgrade was initiated
+ * @property {number} updated_date - timestamp when the log was last updated
+ * @property {string} state - latest upgrade state
+ * @property {Array<Object>} state_history
+ * @property {number} state_history[].date
+ * @property {string} state_history[].state
+ * @property {any} state_history[].details
+ */
+
+const UPGRADE_LOG_STATES = {
+  INITIATED: 'initiated',
+  STAGED: 'staged',
+  INDEXING: 'indexing',
+  INDEXED: 'indexed',
+  COMPLETING: 'completing',
+  FINALIZING: 'finalizing',
+  FINALIZED: 'finalized',
+  COMPLETE: 'complete',
+  ABORTED: 'aborted',
+  ERRORED: 'errored',
+};
+
+let currentUpgradeLogId;
```

Comment:
> I still have concerns about caching this. Is getting the most recent upgrade log from the DB too expensive, or potentially incorrect?

Because it is potentially incorrect. There's also less potential for race conditions, if two users click on the upgrade button at the same time. 

> if we ever decide to run multiple API containers

This is why I used FS before. Would be super nice to have a proper caching service :)

Reply:
I still believe that the local cache saves us from potential race conditions and incorrect DB lookup results. 

Do you think we can bank on having a caching service when we decide to add multiple APIs? We will already need it, because of other things we cache.

Reply:
> This is why I used FS before.

I don't think FS saves us in any way, unless the two containers share a common mount point, right?

> Because it is potentially incorrect. There's also less potential for race conditions, if two users click on the upgrade button at the same time.

I don't follow. With the single container, If two users click go at the same time then two docs will be created and the one that's slightly later will overwrite the `currentUpgradeLogId` in memory right? You'd have two upgrades initiated but only one doc actually being used, and both users would get info from that doc. If instead we relied on the latest doc in the DB, then both users would write a separate doc, the code would only use the latest one, and you'd have the exact same scenario - the slightly later one would be used and the first one would be orphaned.

To be clear, what I'm proposing is instead of storing the ID of the latest doc, we just query the DB for the latest doc. The only benefit I can see with storing it is slightly improved performance (ID lookup rather than view query). The big downside to storing the ID is if some coder forgets to invalidate the cache it'll behave unpredictably.

Reply:
I think it's slightly more correct to keep an upgrade "state" in memory. 
I already have two other places that need it: 
1. the view progress query - that needs to stop on cancel
2. the view indexing queries - that will throw an error if I delete the stated view from under it's feet

For 1, I already added a workaround - by also keeping a reference to the "stop" function in memory. For 2, I switched two calls so the upgrade_log gets the "abort" state before the "errored" state, but the error still gets logged and I'd prefer to catch it. I added a TODO there. 

I'm even inclining to have `prep` immediately create a state object with the necessary bits that gets passed around, and all pending processes can be gracefully handled on a stop. 

Please tell me what you think.